## ひろしまQuest2020#stayhome【コース予測部門】

In [91]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import lightgbm as lgb
import gc
%matplotlib inline
pd.set_option('display.max_Columns', 300)

### 管理番号

In [92]:
model_No = 22

### sub-modelを使用するかどうか

In [93]:
use_sub_model = True

### n/4

In [94]:
sample_No = 4

In [95]:
ALL_MERGE = 'intermediate/{}/all_merge_{}_{}.f'.format(model_No, model_No, sample_No)
SUBMIT = 'submit/course_{}.csv'.format(model_No)
SUBMIT_F = 'submit/course_{}_{}.f'.format(model_No, sample_No)

if use_sub_model:
    ALL_MERGE = 'intermediate/{}/all_merge_{}_{}_sub.f'.format(model_No, model_No, sample_No)
    SUBMIT = 'submit/course_{}_sub.csv'.format(model_No)
else:
    OUT_SUBMODEL = 'intermediate/{}/course_predict_{}_{}.f'.format(model_No, model_No, sample_No)

In [96]:
all_pitch = pd.read_feather(ALL_MERGE)
all_pitch.shape

(778767, 211)

In [97]:
if use_sub_model:
    all_pitch['predict_curve'] = all_pitch['predict_curve'] / all_pitch['predict_straight']
    all_pitch['predict_slider'] = all_pitch['predict_slider'] / all_pitch['predict_straight']
    all_pitch['predict_shoot'] = all_pitch['predict_shoot'] / all_pitch['predict_straight']
    all_pitch['predict_fork'] = all_pitch['predict_fork'] / all_pitch['predict_straight']
    all_pitch['predict_changeup'] = all_pitch['predict_changeup'] / all_pitch['predict_straight']
    all_pitch['predict_sinker'] = all_pitch['predict_sinker'] / all_pitch['predict_straight']
    all_pitch['predict_cutball'] = all_pitch['predict_cutball'] / all_pitch['predict_straight']
    all_pitch.drop(columns=[
        'predict_0', 'predict_1', 'predict_2', 'predict_3', 'predict_4', 'predict_5', 'predict_6',
        'predict_7', 'predict_8', 'predict_9', 'predict_10', 'predict_11', 'predict_12',
       'predict_straight'
    ], inplace=True)

In [98]:
print(all_pitch.shape)
column_cnt = len(all_pitch.columns)
all_pitch.head()

(778767, 197)


No  ball  course  pitch_cnt_in_game  inning  bat_cnt_in_inning  \
0   0   0.0     0.0                  1       1                  1   
1   1   0.0     8.0                  2       1                  1   
2   2   0.0     5.0                  3       1                  1   
3   3   0.0    12.0                  4       1                  2   
4   4   0.0     8.0                  5       1                  2   

   pitch_cnt_in_bat  pitch_order  player_cnt_in_game  pitcher_cnt_in_game  \
0                 1            1                   1                    1   
1                 2            1                   1                    2   
2                 3            1                   1                    3   
3                 1            1                   2                    4   
4                 2            1                   2                    5   

   pitcher_cnt_in_inning  bat_order  bat_cnt_in_game  home_point  away_point  \
0                      1          1                1           0           0   
1                      2          1                1           0           0   
2                      3          1                1           0           0   
3                      4          2                1           0           0   
4                      5          2                1           0           0   

   out_cnt  ball_cnt  strike_cnt  pitch_LR  bat_LR  bc_curve  bc_slider  \
0        0         0           0       0.0     1.0  0.237468   0.325101   
1        0         0           1       0.0     1.0  0.210923   0.247198   
2        0         0           2       0.0     1.0  0.138758   0.130040   
3        1         0           0       0.0     0.0  0.213866   0.474931   
4        1         0           1       0.0     0.0  0.215241   0.420099   

   bc_shoot   bc_fork  bc_changeup  bc_sinker  bc_cutball  bc_high_str  \
0  0.209547  0.145227     0.092978   0.027829    0.113620     0.135883   
1  0.177502  0.320359     0.147952   0.039943    0.116772     0.106283   
2  0.092263  0.419906     0.117327   0.038140    0.046858     0.076007   
3  0.139964  0.088519     0.038543   0.007023    0.138168     0.151308   
4  0.145723  0.196762     0.051566   0.009680    0.118444     0.113468   

   bc_high_ball  bc_mid_str  bc_low_str  bc_low_ball  bc_left_str  \
0      0.221489    0.204959    0.174254     0.263415     0.242045   
1      0.230235    0.157397    0.138195     0.367890     0.184621   
2      0.284249    0.102015    0.095055     0.442674     0.118315   
3      0.196937    0.206965    0.157643     0.287147     0.105057   
4      0.206297    0.162901    0.127988     0.389347     0.084346   

   bc_left_ball  bc_center_str  bc_right_str  bc_right_ball  first  second  \
0      0.294634       0.157681      0.115370       0.190270      0       0   
1      0.355720       0.111331      0.105923       0.242405      0       0   
2      0.488095       0.076190      0.078571       0.238828      0       0   
3      0.153718       0.164756      0.246104       0.330367      0       0   
4      0.162167       0.120564      0.199445       0.433477      0       0   

   third  base_cnt  top_bot  role  pos_pit  date_from_opening  elapsed_min  \
0      0         0        0     1        0                  0     0.000000   
1      0         0        0     1        0                  0     0.200000   
2      0         0        0     1        0                  0     0.383333   
3      0         0        0     1        0                  0     1.016667   
4      0         0        0     1        0                  0     1.200000   

   min_diff  total_inning  elapsed_from_inning  ballnum_from_inning  \
0       NaN             0             0.000000                    1   
1  0.200000             0             0.200000                    2   
2  0.183333             0             0.383333                    3   
3  0.633333             0             1.016667                    4   
4  0.183333             0             1.

### train
- 行数: 257117

In [99]:
train = all_pitch.dropna(subset=['course'])
train.shape

(257117, 197)

### test
- 行数: 521650

In [100]:
test = all_pitch[all_pitch['course'].isnull()]
test.shape

(521650, 197)

In [101]:
del all_pitch
gc.collect()

66

In [102]:
train_d = train.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

In [103]:
test_d = test.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

### LGBM学習
- multiclass : クラス数=8
- tuning結果のパラメータを使用
- 'learning_rate' : 0.1

In [104]:
lgb_param = {
    'objective' : 'multiclass',
    'boosting_type': 'gbdt',
    'metric' : 'multi_logloss',
    'num_class' : 13,
    'seed' : 0,
    'learning_rate' : 0.1,
    'lambda_l1': 8.769293390201968, 
    'lambda_l2': 3.913949617576324e-05, 
    'num_leaves': 6, 
    'feature_fraction': 0.4, 
    'bagging_fraction': 0.8391111798378441, 
    'bagging_freq': 4, 
    'min_child_samples': 50
}

In [105]:
def lightgbm_cv(lgb_param, lgb_train):
    cv_results = lgb.cv(lgb_param, lgb_train,
                    num_boost_round=15000,
                    early_stopping_rounds=100,
                    verbose_eval=100,
                    nfold=4)

    num_boost_round = len(cv_results['multi_logloss-mean'])
    print('Best num_boost_round:', num_boost_round)
    best_cv_score = cv_results['multi_logloss-mean'][-1]
    print('Best CV score:', best_cv_score)
    best_iter = int(num_boost_round * 1.1)
    return best_iter

### Feature Importance

In [106]:
def feature_importance(lgb_model):
    fi = lgb_model.feature_importance()
    fn = lgb_model.feature_name()
    df_feature_importance = pd.DataFrame({'feat_name':fn, 'feat_imp':fi})
    df_feature_importance.sort_values('feat_imp', inplace=True)
    return df_feature_importance

In [107]:
%%time
lgb_train = lgb.Dataset(train_d, train['course'])
best_iter = lightgbm_cv(lgb_param, lgb_train)

[100]	cv_agg's multi_logloss: 2.33586 + 0.000946295
[200]	cv_agg's multi_logloss: 2.32808 + 0.000812927
[300]	cv_agg's multi_logloss: 2.32484 + 0.000783621
[400]	cv_agg's multi_logloss: 2.32306 + 0.000765053
[500]	cv_agg's multi_logloss: 2.32198 + 0.00070057
[600]	cv_agg's multi_logloss: 2.3214 + 0.000736527
[700]	cv_agg's multi_logloss: 2.32094 + 0.000789178
[800]	cv_agg's multi_logloss: 2.32066 + 0.0008135
[900]	cv_agg's multi_logloss: 2.32058 + 0.000848527
[1000]	cv_agg's multi_logloss: 2.32054 + 0.000879605
Best num_boost_round: 972
Best CV score: 2.320525945710633
CPU times: user 1h 45min 5s, sys: 21min 15s, total: 2h 6min 21s
Wall time: 7min 55s


In [108]:
%%time
lgb_model = lgb.train(lgb_param, lgb_train, num_boost_round=best_iter)

CPU times: user 32min 49s, sys: 6min 22s, total: 39min 12s
Wall time: 2min 27s


In [109]:
feature_importance(lgb_model).tail(30)

feat_name  feat_imp
103               slider       640
119            left_ball       665
118             left_str       675
115              mid_str       687
6    pitcher_cnt_in_game       696
122           right_ball       697
140           batter_cnt       709
114            high_ball       729
121            right_str       741
45   elapsed_from_inning       751
116              low_str       755
166          bmi_dif_p-b       774
50    ave_elapsed_batter       778
120           center_str       800
182          bmi_dif_b-c       805
113             high_str       875
53     diff_elapsed_game       937
43              min_diff       938
51   diff_elapsed_batter       953
48      ave_elapsed_game      1006
52   diff_elapsed_inning      1077
49    ave_elapsed_inning      1184
192       predict_sinker      1186
187        predict_curve      1221
193      predict_cutball      1299
191     predict_changeup      1320
188       predict_slider      1342
41     date_from_opening      1585
190         predict_fork      1586
189        predict_shoot      1637

### 予測

In [110]:
%%time
predict = lgb_model.predict(test_d, num_iteration = lgb_model.best_iteration)

CPU times: user 7min 12s, sys: 2.36 s, total: 7min 15s
Wall time: 28.6 s


## 結果出力

In [111]:
submit = pd.DataFrame(predict)
submit.reset_index(inplace=True)
print(submit.shape)
submit.head()

(521650, 14)


index         0         1         2         3         4         5  \
0      0  0.074877  0.139704  0.072007  0.051205  0.057768  0.037280   
1      1  0.067309  0.141530  0.072586  0.038934  0.050440  0.031875   
2      2  0.030005  0.075460  0.051138  0.019822  0.034577  0.023921   
3      3  0.052023  0.111690  0.062563  0.029685  0.033439  0.025583   
4      4  0.047826  0.124583  0.076206  0.033401  0.045125  0.049391   

          6         7         8         9        10        11        12  
0  0.058268  0.050071  0.031850  0.181154  0.102127  0.071596  0.072094  
1  0.035398  0.061821  0.028642  0.205675  0.079346  0.100491  0.085952  
2  0.023026  0.039507  0.021490  0.161566  0.078213  0.307751  0.133524  
3  0.034747  0.053222  0.035278  0.196912  0.102607  0.140370  0.121881  
4  0.041539  0.060587  0.031949  0.169357  0.063031  0.136678  0.120328

### CSV出力

In [112]:
#submit.to_csv(SUBMIT, header=False, index=False)

### Feather出力

In [113]:
submit_f = submit.drop(columns=['index'])
submit_f.rename(columns={
        0: 'predict_0',
        1: 'predict_1',
        2: 'predict_2',
        3: 'predict_3',
        4: 'predict_4',
        5: 'predict_5',
        6: 'predict_6',
        7: 'predict_7',
        8: 'predict_8',
        9: 'predict_9',
        10: 'predict_10',
        11: 'predict_11',
        12: 'predict_12'
    }, inplace=True)
submit_f.to_feather(SUBMIT_F)

### 結果まとめ

In [114]:
ave_result = (sample_No == 4)

In [115]:
if ave_result:
    result = 'submit/course_{}_{}.f'.format(model_No, 1)
    print(result)
    df = pd.read_feather(result)
    columns = ['predict_0', 'predict_1', 'predict_2', 'predict_3', 
               'predict_4', 'predict_5', 'predict_6', 'predict_7',
              'predict_8', 'predict_9', 'predict_10', 'predict_11', 'predict_12']

    for i in range(2,5):
        result = 'submit/course_{}_{}.f'.format(model_No, i)
        print(result)
        temp = pd.read_feather(result)
        for c in columns:
            df[c] = df[c] + temp[c]

    for c in columns:
        df[c] = df[c]/4

    # 出力
    df = df.reset_index()
    df.to_csv(SUBMIT, header=False, index=False)
    print(SUBMIT)

submit/course_22_1.f
submit/course_22_2.f
submit/course_22_3.f
submit/course_22_4.f
submit/course_22_sub.csv


### 球種予測で使用する

In [116]:
%%time
if not use_sub_model:
    train_predict = lgb_model.predict(train_d, num_iteration = lgb_model.best_iteration)

CPU times: user 1 µs, sys: 2 µs, total: 3 µs
Wall time: 7.63 µs


In [117]:
if not use_sub_model:
    df_train_predict = pd.DataFrame(train_predict).reset_index()
    submodel = pd.concat([df_train_predict, submit], ignore_index=True)
    submodel.drop(columns=['index'], inplace=True)
    submodel.rename(columns={
        0: 'predict_0',
        1: 'predict_1',
        2: 'predict_2',
        3: 'predict_3',
        4: 'predict_4',
        5: 'predict_5',
        6: 'predict_6',
        7: 'predict_7',
        8: 'predict_8',
        9: 'predict_9',
        10: 'predict_10',
        11: 'predict_11',
        12: 'predict_12'
    }, inplace=True)
    print(submodel.shape)
    submodel.head()

In [118]:
if not use_sub_model:
    submodel.to_feather(OUT_SUBMODEL)

### SIGNATE

In [119]:
'signate submit --competition-id=276 ./{} --note feat={}'.format(SUBMIT, column_cnt)

'signate submit --competition-id=276 ./submit/course_22_sub.csv --note feat=197'

### GCS

In [120]:
'gsutil cp ./{} gs://yasunobu_baseball/submit/'.format(SUBMIT)

'gsutil cp ./submit/course_22_sub.csv gs://yasunobu_baseball/submit/'

### 評価結果
#### 22_sub(197)
- (1) Best CV score: 2.320717948163731
- (2) Best CV score: 2.3204691031051947
- (3) Best CV score: 2.3217573750471856
- (4) Best CV score: 2.320525945710633
- 評価結果  : 2.3517300

#### 22(190)
- (1) Best CV score: 2.3350431782409524
- (2) est CV score: 2.335302207655757
- (3) Best CV score: 2.3353277079714583
- (4) Best CV score: 2.3340380616284415
- 評価結果  : 2.3487659 ★

#### 21_sub(198)
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 639
- Best CV score: 2.3159487785697754
- Wall time: 5min
- 評価結果  : 2.3572586

#### 20_sub(173)
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 793
- Best CV score: 2.315349281506766
- Wall time: 9min 57s
- 評価結果  : 2.3585163

#### 20(166) ... player同士の組み合わせを削除
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 572
- Best CV score: 2.3271237073894153
- Wall time: 6min 59s
- 評価結果  : 2.3552697

#### 19_sub(197) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 774
- Best CV score: 2.316818574234304
- Wall time: 10min 49s
- 評価結果  : 2.3581029

#### 19(190) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 550
- Best CV score: 2.327580481542456
- Wall time: 7min 36s
- 評価結果  : 2.3554862

#### 18_sub(225) ...サブモデルで球種の予測を追加
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 548
- Best CV score: 2.308116414786115
- Wall time: 8min 25s
- 評価結果  : 2.3638683

#### 18(217) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 373
- Best CV score: 2.3204478805445055
- Wall time: 6min 9s
- 評価結果  : 2.3604751

#### 17(192) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 415
- Best CV score: 2.3200530570975535
- Wall time: 15min 10s
- 評価結果  : 2.3607863

#### 16(181)  ...選手IDごとの2017年のコース実績
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 355
- Best CV score: 2.3203553166835866
- Wall time: 5min 8s
- 評価結果  : 2.3608998

#### 15(153)  ...左右で球種/コースを別に集計
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 932
- Best CV score: 2.3387434142884445
- Wall time: 9min 41s
- 評価結果  : 2.3552543

#### 14(153)  ...左右で球種を別に集計
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 909
- Best CV score: 2.340082741044867
- Wall time: 9min 40s
- 評価結果  : 2.3555548

#### 3(157)  ...前の投球からの時間差を追加
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 857
- Best CV score: 2.340703597440808
- Wall time: 9min 20s
- 評価結果  : 2.3554087

#### 2(156)  ...merge12のパラメータをチューニング
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 835
- Best CV score: 2.3419206656914
- Wall time: 9min 7s
- 評価結果  : 2.3555299

#### 2(156)  ...merge12(2017のコースデータを付加) 
- 'learning_rate' : 0.1,
- Best num_boost_round: 121
- Best CV score: 2.344104226715559
- Wall time: 12min 30s
- 評価結果  : 2.3564326

#### 1(143)  ...球種予測の11と同じデータ 
- 'learning_rate' : 0.1,
- Best num_boost_round: 122
- Best CV score: 2.3445089434874253
- Wall time: 12min 2s (local)
- 評価結果  : 2.3562648